nautobot/core/models/__init__.py

```diff
@@ -1,11 +1,58 @@
 import uuid
+from urllib.parse import quote_plus
 
+from django.contrib.contenttypes.models import ContentType
 from django.db import models
+from natural_keys import NaturalKeyModel, NaturalKeyModelManager
+from natural_keys.models import extract_nested_key
 
 from nautobot.core.models.querysets import RestrictedQuerySet
 
 
-class BaseModel(models.Model):
+class BaseManager(NaturalKeyModelManager):
+    def get_queryset(self):
+        return self._queryset_class(self.model, using=self._db, hints=self._hints)
+
+    def get_by_natural_key(self, *args):
+        """
+        Return the object corresponding to the provided natural key.
+
+        (Extension of django-natural-keys implementation of this generic Django API.)
+        """
+        kwargs = self.natural_key_kwargs(*args)
+
+        # Since kwargs already has __ lookups in it, we could just do "return self.get(**kwargs)"
+        # But django-natural-keys wants to call each related model's get_by_natural_key in case it's overridden:
+        for name, rel_to in self.model.get_natural_key_info():
+            if not rel_to:
+                # Not a related object, no processing needed
+                continue
+
+            # Extract natural key for related object
+            try:
+                nested_key = extract_nested_key(kwargs, rel_to, name)
+            except AttributeError:
+                # Handle case where rel_to isn't a NaturalKeyModel, e.g. ContentType
+                nested_key = []
+                for key in kwargs.keys():
+                    if key.startswith(f"{name}__"):
+                        nested_key.append(kwargs[key])
+                if all(key is None for key in nested_key):
+                    nested_key = None
```

Comment:
There's almost certainly a more elegant way to do this - possibly by reimplementing `extract_nested_key` directly to handle the non-natural-key-model cases within it.

Reply:
All things considered this isn't all that bad.